crates/burn-jit/src/kernel/conv/conv2d/im2col.rs

```diff
@@ -0,0 +1,286 @@
+use burn_tensor::{
+    ops::{conv::calculate_conv_output_size, ConvOptions, FloatTensorOps as _},
+    Shape,
+};
+use cubecl::{calculate_cube_count_elemwise, prelude::*};
+
+use crate::{
+    kernel::into_contiguous,
+    ops::{numeric::empty_device, reshape, swap_dims},
+    tensor::JitTensor,
+    FloatElement, IntElement, JitBackend, JitRuntime,
+};
+
+#[derive(CubeLaunch)]
+struct Im2ColArgs {
+    stride_h: u32,
+    stride_w: u32,
+    dilation_h: u32,
+    dilation_w: u32,
+    padding_h: i32,
+    padding_w: i32,
+
+    kernel_h: u32,
+    kernel_w: u32,
+    out_h: u32,
+    out_w: u32,
+
+    col_size_1: u32,
+    num_elements: u32,
+}
+
+#[cube(launch_unchecked)]
+fn im2col_kernel<F: Float>(
+    image: &Tensor<F>,
+    columns: &mut Tensor<F>,
+    args: &Im2ColArgs,
+    #[comptime] kernel_w_unroll: Option<u32>,
+) {
+    // position shape: [in_channels, batch_size, out_h, out_w]
+    // columns shape: [[in_channels, kernel_h, kernel_w], [batch_size, out_h, out_w]]
+
+    let batch_size = image.shape(0);
+    let height = image.shape(2);
+    let width = image.shape(3);
+
+    let out_h = args.out_h;
+    let out_w = args.out_w;
+
+    if ABSOLUTE_POS > args.num_elements {
+        return;
+    }
+
+    let out_x = ABSOLUTE_POS % out_w;
+    let out_y = ABSOLUTE_POS / out_w % out_h;
+    let batch = ABSOLUTE_POS / (out_w * out_h) % batch_size;
+    let channel = ABSOLUTE_POS / (out_w * out_h * batch_size) % image.shape(1);
+
+    let kernel_w = kernel_w_unroll.unwrap_or(args.kernel_w);
+    let unroll_w = kernel_w_unroll.is_some();
+
+    let image_idx = batch * image.stride(0) + channel * image.stride(1);
+    let col_idx = channel * args.kernel_h * kernel_w * args.col_size_1
+        + batch * out_h * out_w
+        + out_y * out_w
+        + out_x;
+
+    for kernel_y in 0..args.kernel_h {
+        #[unroll(unroll_w)]
+        for kernel_x in 0..kernel_w {
+            let kernel_pos = kernel_y * kernel_w + kernel_x;
+            let col_pos = col_idx + kernel_pos * args.col_size_1;
+
+            let y = (out_y * args.stride_h + kernel_y * args.dilation_h) as i32 - args.padding_h;
+            let x = (out_x * args.stride_w + kernel_x * args.dilation_w) as i32 - args.padding_w;
+            if y >= 0 && x >= 0 && y < height as i32 && x < width as i32 {
```

Comment:
Are there configurations where we are guaranteed to always be within bounds? Is so we could have a comptime strategy to remove the if when not necessary.

Reply:
Since we're iterating over out dimensions it should always be in bounds as long as padding == 0. I'll add a comptime path for that.